### PR TITLE
Modem cmux DCE

### PR DESCRIPTION
**Pull request by jheiskan81**
**Modem cmux update**

- Added missing command and message handling for use existing modem CMUX for DCE role.
- DCE CMUX connection can be now initialized from DTE side.

**Modem test update**

- Added possibility for link 2 modem backend mock instance for validating P2P communication without any data stubbing.
- Added unite test for for validate modem CMUX DTE & DCE role communication together.

### CODE REVIEW

**Review by bjarki-andreasen — CHANGES_REQUESTED** [review]
Really like it, I want to discuss the flow control API, see my comment :)

File: tests/subsys/modem/modem_cmux_pair/src/main.c [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+#define EVENT_CMUX_DLCI1_OPEN    BIT(1)
+#define EVENT_CMUX_DLCI2_OPEN    BIT(2)
+#define EVENT_CMUX_DLCI1_CLOSED  BIT(3)
+#define EVENT_CMUX_DLCI2_CLOSED  BIT(4)
+#define EVENT_CMUX_DISCONNECTED  BIT(5)
+#define EVENT_CMUX_DLCI1_RX_DATA BIT(6)
+#define EVENT_CMUX_DLCI2_RX_DATA BIT(7)
+
+/* CMUX DTE variables */
+static struct modem_cmux cmux;
```
[the remark this answers is not in the document]
Comment (jheiskan81) [reply tests/subsys/modem/modem_cmux_pair/src/main.c:25]: Ah true. I spotted that EVENT_CMUX_CONNECTED was validated twice

**Comment by jheiskan81** [comment]
I update test with your comment's and added some extra verfify for test_modem_cmux_disconnect_connect_sync() and test_modem_cmux_dlci_close_open_sync() test cases for DCE.

File: include/zephyr/modem/cmux.h [hunk 2]
```diff
@@ -270,6 +271,20 @@ int modem_cmux_disconnect(struct modem_cmux *cmux);
...
+ * @details This will control CMUX DLCI data request and and Indication functionality
+ *
+ * @param cmux CMUX instance.
+ * @param state Flow control state true for enable and false for disable.
+ *
+ * @return 0  Opreation OK.
+ * @return -EPIPE CMUX control channel is not open.
+ * @return -EALREADY Ongoing flow control request.
+ */
+int modem_cmux_flow_ctrl_set_asynch(struct modem_cmux *cmux, bool state);
```
[the remark this answers is not in the document]
Comment (bjarki-trackunit) [reply include/zephyr/modem/cmux.h:286]: I'm really not sold on the manual API...

The CMUX instance knows the state of all buffers, its TX buffer, the CMUX RX buffer, and every DLCI RX buffer. CMUX defines two types of flow control, the flow control messages `FCOFF` and `FCON`, and individual channel flow control through `MSC`. We know we need to disable flow control for all channels if the CMUX RX buffer nears full, and we know the same for each DLCI RX buffer, so we really should be handling it automatically from within the CMUX instance.

Flow control should probably be its own PR, or at least its own commit, and should be handled internally. Its the simpler and most scalable solution.

**Review by bjarki-andreasen — CHANGES_REQUESTED** [review]
I would suggest you move flow control out into its own PR, it should be handled internally by CMUX. The DCE additions and test case look excellent :)

**Comment by jheiskan81** [comment]
@bjarki-trackunit Thanks for review. I remove Flow Control API and functionality. I have updated PR.

**Comment by bjarki-trackunit** [comment]
Nice, I will re-review later today

**Comment by jheiskan81** [comment]
@bjarki-trackunit Please review. I have put flow control to original format back.

File: subsys/modem/modem_cmux.c [hunk 3]
```diff
@@ -435,6 +446,40 @@ static void modem_cmux_on_control_frame_uih(struct modem_cmux *cmux)
 	}
 }
 
+static void modem_cmux_conect_response_transmit(struct modem_cmux *cmux)
```
Comment (bjarki-trackunit) [inline subsys/modem/modem_cmux.c:449]: Should be `_connect_`

**Review by bjarki-andreasen — CHANGES_REQUESTED** [review]
Tested the changes locally and it behaves as expected :) There is one spelling mistake, when that is fixed we are good to go :)

File: subsys/modem/modem_cmux.c (hunk 3 above)
Comment (jheiskan81) [reply subsys/modem/modem_cmux.c:449]: I will be fix this after my bicycle ride. Thanks

**Review by bjarki-andreasen — APPROVED** [review]
Nice work :)

**Review by fabiobaltieri — APPROVED** [review]

**Review by rerickson1 — APPROVED** [review]

**Pull request merged**